### PR TITLE
Add general IfEquals(other) method

### LINKED ISSUE #13: Add general .IfEquals(other) method
`Throw().IfEquals(other)` is supported only for objects that implement `IComperable`.

Since all objects have at least a default `Equals` method, we can add a general `.IfEquals(other)` method

### PR DESCRIPTION
**Pull request by DevJonie**
Resolves #13

### CODE REVIEW

**Comment by DevJonie** [comment]
Hi @mantinband,
I wanted to move IfNot/Equals methods to the Equalities files since TValue no longer has the IComparable constraint.
That's if this implementation is Ok. I have a feeling you might not like the branching 😅

**Comment by mantinband** [comment]
> Hi @mantinband, I wanted to move IfNot/Equals methods to the Equalities files since TValue no longer has the IComparable constraint. That's if this implementation is Ok. I have a feeling you might not like the branching in the IsEqual method 😅

Yeah, I don't love the branching. It's another `if` in the way which adds latency. 
I originally thought of adding another `IfEquals` without the `IComperable` constraint.
Then, if the value implements `IComperable` the compiler knows to call the `IComperable` overload, otherwise it will call the general `IfEquals`.
What do you think?

**Comment by DevJonie** [comment]
> Yeah, I don't love the branching. It's another `if` in the way which adds latency. I originally thought of adding another `IfEquals` without the `IComperable` constraint. Then, if the value implements `IComperable` the compiler knows to call the `IComperable` overload, otherwise it will call the general `IfEquals`. What do you think?

I agree with you. In fact, that was my first approach but I realized the types (even primitive types) that implement IComparable don't call the IComparable overload unless it's casted, which is not a natural thing for a developer to do.

- Regular types that Implement IComparable calls the generic overload
```csharp
1.Throw().IfEquals(1);
obj.Throw().IfEquals(otherObj);
///
public static ref readonly Validatable<TValue> IfEquals<TValue>(this in Validatable<TValue> validatable, TValue n)
```

- Type has to be casted before it calls the IComparable overload
```csharp
((IComparable)1).Throw().IfEquals(1);
((IComparable)obj).Throw().IfEquals(otherObj);
///
public static ref readonly Validatable<IComparable> IfEquals(this in Validatable<IComparable> validatable, IComparable n)
```

**Comment by DevJonie** [comment]
> I ran some benchmarks, and turns out that The default comparison (i.e. `EqualityComparer<>.Default..`) is faster, sometimes 100x times faster: image
> 
> So let's change it so we use only the default?

I'm back! Took some time off coding. 😃 

If I understand clearly, you are proposing we remove IComparable from the existing implementation and replace Comparer<>.... with EqualityComparer<>... ?

**Comment by DevJonie** [comment]
Hello @mantinband 

> > I ran some benchmarks, and turns out that The default comparison (i.e. `EqualityComparer<>.Default..`) is faster, sometimes 100x times faster: image
> > So let's change it so we use only the default?
> 
> I'm back! Took some time off coding. 😃
> 
> If I understand clearly, you are proposing we remove IComparable constraint from the existing implementation and replace Comparer<>.... with EqualityComparer<>... ?

Hello @mantinband,
kindly confirm the above so I go on with the implementation.

**Comment by mantinband** [comment]
> Hello @mantinband
> 
> > > I ran some benchmarks, and turns out that The default comparison (i.e. `EqualityComparer<>.Default..`) is faster, sometimes 100x times faster: image
> > > So let's change it so we use only the default?
> > 
> > 
> > I'm back! Took some time off coding. 😃
> > If I understand clearly, you are proposing we remove IComparable constraint from the existing implementation and replace Comparer<>.... with EqualityComparer<>... ?
> 
> Hello @mantinband, kindly confirm the above so I go on with the implementation.

Hey! Sorry, been super busy with work and missed your message. Yes, seems it would be better to always use the `EqualityComparer<>.Default` to check if objects are equal

**Comment by mantinband** [comment]
Looks great, thanks for adding this! 🙏🏽

File: tests/UnitTests/ValidatableExtensions/ValidatableExtensions.EqualitiesTests.cs [hunk 1]
```diff
@@ -58,4 +58,199 @@ public void ThrowIfNotDefault_WhenValueIsDefault_ShouldNotThrow()
...
+        {
+            var other = obj as OverrideEqualsType;
+
+            if (other is null) return false;
+
+            return Id == other.Id;
+        }
+
+        public override int GetHashCode() => Id.GetHashCode();
+    }
```
Comment (mantinband) [inline tests/UnitTests/ValidatableExtensions/ValidatableExtensions.EqualitiesTests.cs:254]: Awesome 👍🏽

**Review by mantinband — APPROVED** [review]

**Pull request merged**